src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs

```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
+        //     Graphics g = Graphics.FromImage(mf); // get a graphics instance
+        //     g.Dispose();  mf.Dispose(); // dispose of the graphics instance first
+        //   OR
+        //     mf.Dispose(); g.Dispose();  // dispose of the metafile instance first
+        //
+        // ligbgdiplus has a bug where disposing of the metafile instance first will
+        // trigger a use of freed memory when the graphics instance is disposed, which
+        // could lead to crashes when the native memory is reused.
+        //
+        // The metafile holder is designed to take ownership of the native metafile image
+        // when the managed Metafile instance is disposed while a Graphics instance is still
+        // not disposed (ie the second code pattern above) and to keep the native image alive until the graphics
+        // instance is disposed.
+        //
+        // Note that the following throws, so we only ever need to keep track of one Graphics
+        // instance at a time:
+        //      Metafile mf = ...;  // get a metafile instance
+        //      Graphics g = Graphics.FromImage(mf);
+        //      Graphics g2 = Graphics.FromImage(mf);  // throws OutOfMemoryException on GDI+ on Win32
+        internal sealed class MetafileHolder : IDisposable
+        {
+            private bool disposed;
+            private IntPtr _nativeImage;
+
+
+            internal bool Disposed { get => disposed; }
+            internal MetafileHolder ()
+            {
+                disposed = false;
+                _nativeImage = IntPtr.Zero;
+            }
+
+            ~MetafileHolder () => Dispose(false);
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            internal void Dispose(bool disposing)
+            {
+                if (disposed)
+                    return;
+                IntPtr nativeImage = _nativeImage;
+                _nativeImage = IntPtr.Zero;
+                disposed = true;
+                if (nativeImage != IntPtr.Zero)
+                    Metafile.DisposeNativeInstance (nativeImage);
+            }
+
+            internal void MetafileDisposed (IntPtr nativeImage)
+            {
+                _nativeImage = nativeImage;
+            }
+
+            internal void GraphicsDisposed ()
+            {
+                Dispose();
+            }
+        }
+
+        internal MetafileHolder? AddMetafileHolder(Graphics g)
+        {
+            // If _metafileHolder is not null and hasn't been disposed yet, there's already a graphics instance associated with
+            // this metafile, the native code will return an error status.
+            if (_metafileHolder != null && !_metafileHolder.Disposed)
+                return null;
+            _metafileHolder = new MetafileHolder();
+            return _metafileHolder;
+        }
+
+        internal static void DisposeNativeInstance(IntPtr nativeImage)
```

Comment:
Why make this a method if it is only used in one place? Should we just put this inline?